modules/imap/hm-imap.php

```diff
@@ -233,40 +233,143 @@ public function disconnect() {
                 fclose($this->handle);
             }
         }
-
+     // hashes array for scram authentication login case
+        private $hashes = array(
+            'sha-1'   => 'sha1',
+            'sha1'    => 'sha1',
+            'sha-224' => 'sha224',
+            'sha224'  => 'sha224',
+            'sha-256' => 'sha256',
+            'sha256'  => 'sha256',
+            'sha-384' => 'sha384',
+            'sha384'  => 'sha384',
+            'sha-512' => 'sha512',
+            'sha512'  => 'sha512'
+        );
+        
+        private function getHashAlgorithm($scramAlgorithm) {
+            $parts = explode('-', strtolower($scramAlgorithm));
+            if (isset($this->hashes[$parts[1]])) {
+                return $this->hashes[$parts[1]];
+            }
+            return 'sha1'; // Default to sha1 if the algorithm is not found
+        }
+        
+        // Function to generate client proof in SCRAM mechanism
+        private function generateClientProof($username, $password, $salt, $clientNonce, $serverNonce, $algorithm) {
+            $iterations = 4096;
+            $keyLength = strlen(hash($algorithm, '', true)); // Dynamically determine key length based on algorithm
+        
+            $passwordBytes = hash($algorithm, $password, true);
+            $saltedPassword = hash_pbkdf2($algorithm, $passwordBytes, $salt, $iterations, $keyLength, true);
+            $clientKey = hash_hmac($algorithm, "Client Key", $saltedPassword, true);
+            $storedKey = hash($algorithm, $clientKey, true);
+            $authMessage = 'n=' . $username . ',r=' . $clientNonce . ',s=' . base64_encode($salt) . ',r=' . $serverNonce;
+            $clientSignature = hash_hmac($algorithm, $authMessage, $storedKey, true);
+            $clientProof = base64_encode($clientKey ^ $clientSignature);
+            return $clientProof;
+        }
+        
+        private function authenticateScram($scramAlgorithm, $username, $password) {
+            $algorithm = $this->getHashAlgorithm($scramAlgorithm);
+            
+            // Send initial SCRAM command
+            $scramCommand = 'AUTHENTICATE ' . $scramAlgorithm . "\r\n";
+            $this->send_command($scramCommand);
+            $response = $this->get_response();
+            if (!empty($response) && substr($response[0], 0, 2) == '+ ') {
```

Comment:
Can you debug/log messages coming from the server that are not in the right/expected format to make it easier for people to debug problems with client/server communication later?

Reply:
alright let me do that